RecoTracker/MkFitCMS/src/MkStdSeqs.cc

```diff
@@ -586,39 +594,131 @@ namespace mkfit {
+    //   this implicitly separates triplets and doublet seeds with glued layers
+    template <class TRACK>
+    bool qfilter_n_hits(const TRACK &t, const MkJob &j) {
+      int seedHits = t.getNSeedHits();
+      int seedReduction = (seedHits <= 5) ? 2 : 3;
+      return t.nFoundHits() - seedReduction >= j.params_cur().minHitsQF;
+    }
 
-#ifdef DEBUG
-      std::cout << " find_and_remove_duplicates: output track size " << tracks.size() << std::endl;
-      for (auto const &tk : tracks) {
-        std::cout << tk.parameters() << std::endl;
-      }
-#endif
+    // simple hit-count quality filter; used with pixel-based seeds
+    template <class TRACK>
+    bool qfilter_n_hits_pixseed(const TRACK &t, const MkJob &j) {
+      return t.nFoundHits() >= j.params_cur().minHitsQF;
+    }
+
+    // layer-dependent quality filter
+    template <class TRACK>
+    bool qfilter_n_layers(const TRACK &t, const MkJob &j) {
+      const BeamSpot &bspot = j.m_beam_spot;
+      const TrackerInfo &trk_inf = j.m_trk_info;
+      int enhits = t.nHitsByTypeEncoded(trk_inf);
+      int npixhits = t.nPixelDecoded(enhits);
+      int enlyrs = t.nLayersByTypeEncoded(trk_inf);
+      int npixlyrs = t.nPixelDecoded(enlyrs);
+      int nmatlyrs = t.nTotMatchDecoded(enlyrs);
+      int llyr = t.getLastFoundHitLyr();
+      int lplyr = t.getLastFoundPixelHitLyr();
+      float invpt = t.invpT();
+      float invptmin = 1.43;  // min 1/pT (=1/0.7) for full filter on (npixhits<=3 .or. npixlyrs<=3)
+      float d0BS = t.d0BeamSpot(bspot.x, bspot.y);
+      float d0_max = 0.1;  // 1 mm
+
+      bool endsInsidePix = (llyr == 2 || llyr == 18 || llyr == 45);
+      bool lastInsidePix = ((0 <= lplyr && lplyr < 3) || (18 <= lplyr && lplyr < 20) || (45 <= lplyr && lplyr < 47));
+      return !(((npixhits <= 3 || npixlyrs <= 3) && endsInsidePix &&
+                (invpt < invptmin || (invpt >= invptmin && std::abs(d0BS) > d0_max))) ||
+               ((npixlyrs <= 3 && nmatlyrs <= 6) && lastInsidePix && llyr != lplyr && std::abs(d0BS) > d0_max));
     }
 
+    /// quality filter tuned for pixelLess iteration during forward search
+    template <class TRACK>
+    bool qfilter_pixelLessFwd(const TRACK &t, const MkJob &j) {
+      const BeamSpot &bspot = j.m_beam_spot;
+      const TrackerInfo &tk_info = j.m_trk_info;
+      float d0BS = t.d0BeamSpot(bspot.x, bspot.y);
+      float d0_max = 0.05;  // 0.5 mm
+
+      int encoded;
+      encoded = t.nLayersByTypeEncoded(tk_info);
+      int nLyrs = t.nTotMatchDecoded(encoded);
+      encoded = t.nHitsByTypeEncoded(tk_info);
+      int nHits = t.nTotMatchDecoded(encoded);
+
+      int seedReduction = (t.getNSeedHits() <= 5) ? 2 : 3;
+
+      float invpt = t.invpT();
+      float invptmin = 1.11;  // =1/0.9
+
+      float thetasym = std::abs(t.theta() - Const::PIOver2);
+      float thetasymmin = 1.11;  // -> |eta|=1.45
+
+      return (((t.nFoundHits() - seedReduction >= 4 && invpt < invptmin) ||
```

Comment:
More magic numbers. Should they go into a config or from a DB?

Reply:
this is old code, not a part of this PR (just a relocation in this PR).

it will become configurable when a dynamic change is needed; compiled constants are preferred until then for performance reasons.

The values are ad-hoc and making them named constants is not practical (to avoid `int four = 4;`)

Reply:
actually, my response should've been at the level at the function: it is already configurable, but at the level of the function call, as the purpose of this PR. If a different selection is derived, it will be encoded as another function that can be changed in the corresponding iteration json file.